TAC-4143 | acquia-platform-cloud-data-model needs a DbBackup class.

kirbpowell opened the pull request:
This commit adds `DbBackupInterface.php`, `DbBackup.php` & `DbBackupDecoratorMethods.php`, along with associated unit tests to the Cloud Platform Data Model. 

These classes and the methods within them take after the example of `DbInstance`.

kirbpowell committed:
TAC-4143 | acquia-platform-cloud-data-model needs a DbBackup class.

pbull commented on tests/Hosting/DbBackup/DbBackupDecoratorMethodsTest.php:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    /**
+     * @covers ::setId
```
Unit test failures on this @covers statement - setId is not a valid method.

```
1.87s$ vendor/bin/phpunit

PHPUnit 4.8.26 by Sebastian Bergmann and contributors.

...............................................................  63 / 390 ( 16%)

.............................Trying to @cover or @use not existing method "\Acquia\Platform\Cloud\Hosting\DbBackup\DbBackupDecoratorMethods::setId".
```

pbull commented:
PHP Codesniffer checks that failed under travis. Note that the ServerTest failures were due to an update of the version of phpcs running and not related to your PR. I'll push up a fix for that shortly:

```
$ vendor/bin/phpcs -s --report-width=120 --standard=PSR2 ./{src,tests}

FILE: /home/travis/build/webbj74/acquia-platform-cloud-data-model/src/Hosting/DbBackup.php

------------------------------------------------------------------------------------------------------------------------

FOUND 1 ERROR AFFECTING 1 LINE

------------------------------------------------------------------------------------------------------------------------

 12 | ERROR | [x] There must be one blank line after the namespace declaration

    |       |     (PSR2.Namespaces.NamespaceDeclaration.BlankLineAfter)

------------------------------------------------------------------------------------------------------------------------

PHPCBF CAN FIX THE 1 MARKED SNIFF VIOLATIONS AUTOMATICALLY

------------------------------------------------------------------------------------------------------------------------

FILE: /home/travis/build/webbj74/acquia-platform-cloud-data-model/src/Hosting/DbBackupInterface.php

----------------------------------------------------------------------------------------------------

FOUND 1 ERROR AFFECTING 1 LINE

----------------------------------------------------------------------------------------------------

 156 | ERROR | [x] Expected 1 newline at end of file; 0 found (PSR2.Files.EndFileNewline.NoneFound)

----------------------------------------------------------------------------------------------------

PHPCBF CAN FIX THE 1 MARKED SNIFF VIOLATIONS AUTOMATICALLY

----------------------------------------------------------------------------------------------------

FILE: /home/travis/build/webbj74/acquia-platform-cloud-data-model/tests/Hosting/ServerTest.php

------------------------------------------------------------------------------------------------------------------------

FOUND 3 ERRORS AFFECTING 3 LINES

------------------------------------------------------------------------------------------------------------------------

 167 | ERROR | [x] Function closing brace must go on the next line following the body; found 1 blank lines before brace

     |       |     (PSR2.Methods.FunctionClosingBrace.SpacingBeforeClose)

 210 | ERROR | [x] Function closing brace must go on the next line following the body; found 1 blank lines before brace

     |       |     (PSR2.Methods.FunctionClosingBrace.SpacingBeforeClose)

 253 | ERROR | [x] Function closing brace must go on the next line following the body; found 1 blank lines before brace

     |       |     (PSR2.Methods.FunctionClosingBrace.SpacingBeforeClose)

------------------------------------------------------------------------------------------------------------------------

PHPCBF CAN FIX THE 3 MARKED SNIFF VIOLATIONS AUTOMATICALLY

------------------------------------------------------------------------------------------------------------------------

FILE: /home/travis/build/webbj74/acquia-platform-cloud-data-model/tests/Hosting/DbBackupTest.php

------------------------------------------------------------------------------------------------------------------------

FOUND 2 ERRORS AFFECTING 2 LINES

------------------------------------------------------------------------------------------------------------------------

  14 | ERROR | [x] There must be one blank line after the last USE statement; 2 found;

     |       |     (PSR2.Namespaces.UseDeclaration.SpaceAfterLastUse)

 351 | ERROR | [x] Expected 1 newline at end of file; 0 found (PSR2.Files.EndFileNewline.NoneFound)

------------------------------------------------------------------------------------------------------------------------

PHPCBF CAN FIX THE 2 MARKED SNIFF VIOLATIONS AUTOMATICALLY

------------------------------------------------------------------------------------------------------------------------
```

pbull commented:
@kirbpowell https://github.com/webbj74/acquia-platform-cloud-data-model/pull/34 is a really quick PR that will unblock the test failures in ServerTest.

kirbpowell commented:
Just ran PHPCBF on all the above failures and got 

```
No fixable errors were found
```

for all of the files. A manual inspection appeared to show necessary end-of-file newlines and other blankspaces, so hopefully they're fixed now. Did fix the travis test that failed outright.

kirbpowell commented:
Fixed all failing tests, and fixed all phpcs errors. The remaining failures deal with complexity issues in the new classes, and I'm not sure how to avoid them.

webbj74 commented on src/Hosting/DbBackup.php:
```diff
@@ -0,0 +1,324 @@
+                )
+            );
+        }
+        $this->id = $backupId;
```
Shouldn't this be $this->backupId?

webbj74 commented on src/Hosting/DbBackup.php:
```diff
@@ -0,0 +1,324 @@
+        }
+        if (isset($dbBackupData['checksum'])) {
+            $app->setChecksum($dbBackupData['checksum']);
+        }
```
See https://github.com/webbj74/acquia-platform-cloud-data-model/blob/master/src/Hosting/Application.php#L87-L100 for less complex way to do this

kirbpowell commented:
Updated the commit to reflect @webbj74 's comments and now everything is passing! Should be good to go

pbull commented on src/Hosting/DbBackup.php:
```diff
@@ -0,0 +1,341 @@
+    public static function create(array $dbBackupData)
+    {
+        $app = new static($dbBackupData['id']);
+//        if (isset($dbBackupData['type'])) {
```
We can ditch the commented out code.

pbull commented on src/Hosting/DbBackupInterface.php:
```diff
@@ -0,0 +1,156 @@
+
+/**
+ * A DbBackup corresponds to a backup of a database for a client's application.
+ * It contains:
```
@kirbpowell This interface is structured after the internal data model... it should reflect the fields returned by the public API (http://cloudapi.acquia.com/#GET__sites__site_envs__env_dbs__db_backups-instance_route).

Specifically:
- the site ID and db ID fields are not returned by the API.
- the public API does return 'name' and  'link' fields.

pbull commented on src/Hosting/DbBackupInterface.php:
```diff
@@ -0,0 +1,156 @@
+ * A DbBackup corresponds to a backup of a database for a client's application.
+ * It contains:
+ * - A backup ID eg. 12345678
+ * - A backup type, eg. daily, weekly, etc
```
Backup types are: daily, ondemand

kirbpowell commented:
This version makes the above changes, and implements only the public API
